client/remoterepo.c

```diff
@@ -106,11 +206,409 @@ _handle_curl_cb(
+        dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    dwError = TDNFCheckHash(pszFile, ml_file->digest, ml_file->type);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+cleanup:
+    if(ml_file)
+    {
+        TDNF_SAFE_FREE_MEMORY(ml_file->digest);
+        TDNF_SAFE_FREE_MEMORY(ml_file->filename);
+        TDNF_SAFE_FREE_MEMORY(ml_file);
+    }
+    return dwError;
+error:
+    goto cleanup;
+}
+
+/* Returns nonzero if hex_digest is properly formatted; that is each
+   letter is in [0-9A-Za-z] and the length of the string equals to the
+   result length of digest * 2. */
+uint32_t
+TDNF_Check_Hex_Digest(
+    const char *hex_digest,
+    int digest_length
+    )
+{
+    int i = 0;
+    if(IsNullOrEmptyString(hex_digest) ||
+       !digest_length)
+    {
+        return 0;
+    }
+    for(i = 0; hex_digest[i]; ++i)
+    {
+        if(!isxdigit(hex_digest[i]))
+        {
+            return 0;
+        }
+    }
+    return digest_length * 2 == i;
+}
+
+uint32_t
+TDNF_HexToUint(
+    const char *hex_digest,
+    unsigned char *return_value
+    )
+{
+    uint32_t dwError = 0;
+    char buf[3] = {0};
+    unsigned long val = 0;
+
+    if(IsNullOrEmptyString(hex_digest) ||
+       !return_value)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    buf[0] = hex_digest[0];
+    buf[1] = hex_digest[1];
+
+    errno = 0;
+    val = strtoul(buf, NULL, 16);
+    if(errno)
+    {
+        fprintf(stderr, "Error: strtoul call failed\n");
+
+        dwError = errno;
+        BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
+    }
+    *return_value = (unsigned char)(val&0xff);
+
+cleanup:
+    return dwError;
+error:
+    goto cleanup;
+}
+
+uint32_t
+TDNF_Checksum_From_Hex_Digest(
+    const char *hex_digest,
+    unsigned char **ppdigest
+    )
+{
+    uint32_t dwError = 0;
+    unsigned char *pdigest = NULL;
+    size_t i = 0;
+    size_t len = 0;
+    unsigned char ret = 0;
+
+    if(IsNullOrEmptyString(hex_digest) ||
+       !ppdigest)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    len = strlen(hex_digest);
+
+    dwError = TDNFAllocateMemory(1, len/2, (void **)&pdigest);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    for(i = 0; i < len; i += 2)
+    {
+        dwError = TDNF_HexToUint(hex_digest + i, &ret);
+        BAIL_ON_TDNF_ERROR(dwError);
+
+        pdigest[i>>1] = ret;
+    }
+    *ppdigest = pdigest;
+
+cleanup:
+    return dwError;
+
+error:
+    TDNF_SAFE_FREE_MEMORY(pdigest);
+    goto cleanup;
+}
+
+int
+TDNF_Get_Resource_Type(
+    const char *resource_type,
+    int  *type
+    )
+{
+    uint32_t dwError = 0;
+
+    if(IsNullOrEmptyString(resource_type) ||
+       !type)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if(!strcasecmp(resource_type, "sha256") || !strcasecmp(resource_type, "sha-256"))
```

Comment:
Can we have a 1-1 mapping for these hash methods and types by using a struct? You've declared an enum already,
So we can do it inside a loop instead of an if else ladder.
I suggest putting everything in _hash_op structure and using it everywhere.

Reply:
if we add that, it will be again a double loop.. trying to avoid it. Also beside "sha1" and "sha1-1" there is no other standard representation.